File: binary-search-tree/binary_search_tree.rb
class Node
  attr_accessor :data, :left, :right
  def initialize(data)
    @data, @left, @right = data, nil, nil
  end
end

class BST
  def initialize
    @root = nil
  end
  
  def insert(data)
    @root = insert_node(@root, data)
  end
  
  def insert_node(node, data)
    return Node.new(data) unless node
    if data < node.data
      node.left = insert_node(node.left, data)
    else
      node.right = insert_node(node.right, data)
    end
    node
  end
  
  def search(data)
    search_node(@root, data)
  end
  
  def search_node(node, data)
    return node if !node || node.data == data
    data < node.data ? search_node(node.left, data) : search_node(node.right, data)
  end
end

bst = BST.new
[50, 30, 70, 20, 40].each { |x| bst.insert(x) }
puts bst.search(40) ? "Found" : "Not found" 

File: binary-search-tree/binary_search_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None
    
    def insert(self, data):
        self.root = self._insert(self.root, data)
    
    def _insert(self, node, data):
        if not node:
            return Node(data)
        
        if data < node.data:
            node.left = self._insert(node.left, data)
        else:
            node.right = self._insert(node.right, data)
        
        return node
    
    def search(self, data):
        return self._search(self.root, data)
    
    def _search(self, node, data):
        if not node or node.data == data:
            return node is not None
        
        if data < node.data:
            return self._search(node.left, data)
        return self._search(node.right, data)

# Test
bst = BST()
bst.insert(50)
bst.insert(30)
bst.insert(70)
print(bst.search(30))  # True
